File: MovieTicketBooking/Services/SeatAvailabilityService.cs
using MovieTicketBooking.Model;
using MovieTicketBooking.Providers;

namespace MovieTicketBooking.Services
{
    public class SeatAvailabilityService
    {
        ShowService _showService;
        ISeatLockProvider _seatLockProvider;
        BookingService _bookingService;
        public SeatAvailabilityService(ShowService showService, ISeatLockProvider seatLockProvider,
        BookingService bookingService)
        {
            _showService = showService;
            _seatLockProvider = seatLockProvider;
            _bookingService = bookingService;
        }
        public List<Seat> GetAllAvailableSeats(string showId)
        {
            Show show = _showService.GetShow(showId);
            Screen screen = show.Screen;
            List<Seat> allSeats = screen.Seats;
            var lockedSeats = GetLockedSeats(show);
            foreach (Seat seat in allSeats)
            {
                if (lockedSeats.Contains(seat))
                {
                    allSeats.Remove(seat);
                }
            }
            return allSeats;
        }
        private List<Seat> GetLockedSeats(Show show)
        {
            List<Seat> seats = new();
            seats.AddRange(_seatLockProvider.GetLockedSeats(show));
            seats.AddRange(_bookingService.GetBookedSeats(show));
            return seats;
        }
    }
}

File: MovieTicketBooking/Movie Booking Java/src/main/java/org/example/service/SeatAvailabilityService.java
package org.example.service;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import org.example.model.Seat;
import org.example.model.Show;
import org.example.providers.SeatLockProvider;

import java.util.List;

@AllArgsConstructor
public class SeatAvailabilityService {

    private final BookingService bookingService;
    private final SeatLockProvider seatLockProvider;

    public List<Seat> getAvailableSeats(@NonNull final Show show) {

        List<Seat> allSeats = show.getScreen().getSeats();
        List<Seat> unAvailableSeats = getUnAvailableSeats(show);

        allSeats.removeAll(unAvailableSeats);
        return allSeats;
    }

    public List<Seat> getUnAvailableSeats(@NonNull final Show show) {

        List<Seat> seats = this.bookingService.getBookedSeats(show);
        seats.addAll(this.seatLockProvider.getLockedSeats(show));
        return seats;
    }
}
